### PR TITLE
Module refactor + Concurrency Fixes #6

### LINKED ISSUE #6: Add concurrency to modules
Multiple commands need to be able to be executed simultaneously.

### PR DESCRIPTION
**Pull request by TheThirdOne**

### CODE REVIEW

File: baton/hub.go [hunk 1]
```diff
@@ -40,45 +40,57 @@ func (h hub) run() {
 				close(c.send)
 			}
 		case rawMsg := <-h.receive:
-			fmt.Println(rawMsg)
 			var cmd command
 			if err := json.Unmarshal(rawMsg.data, &cmd); err != nil {
 				fmt.Println("nu")
```
Comment (zachlatta) [inline baton/hub.go:45]: Do we need this line? (Pretty sure I'm the one that added it though :-P)

File: baton/hub.go [hunk 2]
```diff
@@ -40,45 +40,57 @@ func (h hub) run() {
...
 				log.Println(err)
```
Comment (zachlatta) [inline baton/hub.go:46]: Perhaps we could change this to the following to be a bit more descriptive?

``` go
log.Println("error unmarshaling incoming message:", err)
```

File: baton/hub.go [hunk 3]
```diff
@@ -40,45 +40,57 @@ func (h hub) run() {
...
 				break
 			}
-
+			fmt.Println(cmd)
```
Comment (zachlatta) [inline baton/hub.go:49]: Perhaps make this a log statement instead and add context for the line? Might make sense to use a package like https://github.com/Sirupsen/logrus so we can have logging levels and disable certain levels.

``` go
log.Println("Incoming message/command:", cmd)
```

File: baton/hub.go [hunk 4]
```diff
@@ -40,45 +40,57 @@ func (h hub) run() {
...
+}
 
-			fmt.Println(cmd)
-			resp := make(chan interface{})
-			module, ok := modules[msg.cmd.Module]
+func processMessage(message msg) {
+	resp := make(chan interface{})
+	module, ok := modules[message.cmd.Module]
+	if !ok {
+		fmt.Println(message.cmd.Module, "not in", modules)
```
Comment (zachlatta) [inline baton/hub.go:60]: What if we used a log statement so we also get a timestamp?

Comment (TheThirdOne) [inline baton/hub.go:60]: All of the issues about logging should be ignored until I begin on #12. I will change pretty much all of them then.

File: baton/hub.go (hunk 1 above)
Comment (TheThirdOne) [inline baton/hub.go:45]: That line spits out a bunch of numbers; nothing any human can read.

File: baton/hub.go [hunk 5]
```diff
@@ -40,45 +40,57 @@ func (h hub) run() {
...
+		return
+	}
+
+	go func() {
+		for {
+			r, ok := <-resp
 			if !ok {
-				fmt.Println(msg.cmd.Module, "not in", modules)
+				close(resp)
 				break
```
Comment (zachlatta) [inline baton/hub.go:69]: What will happen after the channel is closed and the for loop (starting on line 65) continues to run?

File: baton/hub.go [hunk 6]
```diff
@@ -40,45 +40,57 @@ func (h hub) run() {
...
+			bytes, err := json.Marshal(command{message.cmd.Module, message.cmd.Call, r}) //add in Module and Call info for client
+			if err != nil {
+				log.Println("Error Marshaling")
+				log.Println(err)
 				break
 			}
+			message.conn.send <- bytes
 		}
-	}
+	}()
```
Comment (zachlatta) [inline baton/hub.go:79]: Will a new goroutine be created and run infinitely on every command? Is this a memory leak?

File: baton/hub.go (hunk 5 above)
Comment (zachlatta) [inline baton/hub.go:69]: Ah, my bad. The `break` exits the for loop. Perhaps `return` might be a bit more clear?

File: baton/hub.go (hunk 1 above)
Comment (zachlatta) [inline baton/hub.go:45]: Yeah, we need to cast the []byte slice to string to get readable output, as so:

``` go
fmt.Println(string(rawMsg))
```

Should we just go ahead and remove it for now?

File: baton/hub.go (hunk 6 above)
Comment (TheThirdOne) [inline baton/hub.go:79]: A new goroutine will be created but will not run infinitely unless the command never ends.

File: baton/hub.go (hunk 1 above)
Comment (TheThirdOne) [inline baton/hub.go:45]: I will probably log it when proper logging is setup. (#12)

File: baton/hub.go (hunk 5 above)
Comment (TheThirdOne) [inline baton/hub.go:69]: Both `break` and `return` are clear to me. If `return` is clearer to you, I'll change it.

**Pull request merged**